### PR TITLE
Add Python implementation of genotype_matrix

### PR DESCRIPTION
**Pull request by szhan**
Part of #678

### CODE REVIEW

File: python/tskit/trees.py [hunk 1]
```diff
@@ -5256,10 +5256,22 @@ def genotype_matrix(
...
 
-        return self._ll_tree_sequence.get_genotype_matrix(
-            isolated_as_missing=isolated_as_missing, alleles=alleles
+        variant = tskit.Variant(
+            self,
+            isolated_as_missing=isolated_as_missing,
+            alleles=alleles,
         )
 
+        ret = np.zeros(shape=(self.num_sites, self.num_samples))
```
Comment (benjeffery) [inline python/tskit/trees.py:5265]:
```suggestion
        ret = np.zeros(shape=(self.num_sites, self.num_samples), dtype=np.int32)
```

**Review by benjeffery — COMMENTED** [review]
Screenshot from 2022-08-16 21-35-55

3x slower than C!



I'll have a think about how to improve this - worst case we use the fast path for when the `samples`/`sites` are something like 60-80% full and use the python way when we are skipping a lot, or just add the arguments in C....

**Comment by jeromekelleher** [comment]
I'm surprised this is so much slower - where is the time being spent?

**Comment by benjeffery** [comment]
> I'm surprised this is so much slower - where is the time being spent?

@szhan Do you want to pair on some perf analysis?

**Comment by szhan** [comment]
C implementation.
```
ts.genotype_matrix();'tiny.trees'  2.0 µ
ts.genotype_matrix();'bench.trees' 287.3 m
```

Python implementation.
```
ts.genotype_matrix();'tiny.trees'  6.2 µ
ts.genotype_matrix();'bench.trees' 455.9 m
```

It is taking 1.6 times longer to run `bench.trees`.

**Comment by jeromekelleher** [comment]
Seems like an acceptable perf tradeoff to me.

We shouldn't forget to remove the Python C implementation, now it's unused.

**Comment by benjeffery** [comment]
@szhan Do you know what to remove in `_tskitmodule.c`, or would you like some pointers? It should be part of this PR.

**Comment by szhan** [comment]
I think `TreeSequence_get_genotype_matrix()` and the snippet `.ml_name = "get_genotype_matrix"`?

**Comment by benjeffery** [comment]
> I think `TreeSequence_get_genotype_matrix()` and the snippet `.ml_name = "get_genotype_matrix"`?

Looks good, just the test that needs removing also.

**Review by benjeffery — APPROVED** [review]
Great, thanks @szhan

**Pull request merged**